ipc_types/types.go

```diff
@@ -4,185 +4,192 @@ package ipc_types
-	Killed
+	RootStr          = "/root"
+	SubnetSeparator  = "/"
+	UndefStr         = ""
+	IPCAddrSeparator = "-"
 )
 
-type ChainEpoch int64
-
-type Validator struct {
-	Addr    address.Address
-	NetAddr string
-}
-
-type ConstructParams struct {
-	Parent            SubnetID
-	Name              string
-	IpcGatewayAddr    uint64
-	Consensus         ConsensusType
-	MinValidatorStake TokenAmount
-	MinValidators     uint64
-	FinalityThreshold ChainEpoch
-	CheckPeriod       ChainEpoch
-	Genesis           []byte
-}
-
-type JoinParams struct {
-	ValidatorNetAddr string
-}
-
-type Votes struct {
-	Validators []address.Address
-}
-
-type Checkpoint struct {
-	Data CheckData
-	Sig  []byte
-}
-
-type CheckData struct {
-	Source    SubnetID
-	TipSet    []byte
-	Epoch     ChainEpoch
-	PrevCheck cid.Cid //TCid<TLink<Checkpoint>>
-	Children  []ChildCheck
-	CrossMsgs []CrossMsgMeta
-}
-
-type ChildCheck struct {
-	Source SubnetID
-	Checks cid.Cid //Vec<TCid<TLink<Checkpoint>>>,
-}
-
-type CrossMsgMeta struct {
-	From    SubnetID
-	To      SubnetID
-	MsgsCID cid.Cid //TCid<TLink<CrossMsgs>>,
-	Nonce   uint64
-	Value   TokenAmount
-}
-
-// ----------------------------------------------------
-// TODO Move them in separate files that somewhat follow the division in Rust
-// TODO here below  ipc-gateways state and params (and other types associated to them)(separate further?)
-// ------------------------------------------------------
-type IPCGatewayState struct { //TODO change name to state once in its own file
-	NetworkName          SubnetID
-	TotalSubnets         uint64
-	MinStake             TokenAmount
-	Subnets              cid.Cid //TCid<THamt<Cid, Subnet>>
-	CheckPeriod          ChainEpoch
-	Checkpoints          cid.Cid //TCid<THamt<ChainEpoch, Checkpoint>>
-	CheckMsgRegistry     cid.Cid //TCid<THamt<TCid<TLink<CrossMsgs>>, CrossMsgs>>
-	Nonce                uint64
-	BottomupNonce        uint64
-	BottomupMsgMeta      cid.Cid //TCid<TAmt<CrossMsgMeta, CROSSMSG_AMT_BITWIDTH>>
-	AppliedBottomupNonce uint64
-	AppliedTopdownNonce  uint64
-}
-
-type ConstructorParams struct {
-	NetworkName      string
-	CheckpointPeriod ChainEpoch
-}
-
-type StorableMsg struct {
-	From   IPCAddress
-	To     IPCAddress
-	Method MethodNum
-	Params RawBytes
-	Value  TokenAmount
-	Nonce  uint64
-}
-
-type MethodNum uint64
-
-type RawBytes struct {
-	Bytes []byte
-}
-
-type CrossMsg struct {
-	Msg     StorableMsg
-	Wrapped bool
-}
-
-type FundParams struct {
-	Value TokenAmount
-}
-
-type CrossMsgParams struct {
-	CrossMsg    CrossMsg
-	Destination SubnetID
-}
-
-type ApplyMsgParams struct {
-	CrossMsg CrossMsg
-}
-
-type CrossMsgs struct {
-	Msgs  []CrossMsg
-	Metas []CrossMsgMeta
+// RootSubnet is the ID of the root network
+var RootSubnet = SubnetID{
+	Parent: RootStr,
+	Actor:  id0,
+}
+
+// UndefSubnetID is the undef ID
+var UndefSubnetID = SubnetID{
+	Parent: UndefStr,
+	Actor:  id0,
+}
+
+func NewSubnetIDFromString(parent string, subnetAct address.Address) *SubnetID {
```

Comment:
This one is wrong `NewSubnetIDFromString` should expect a `string` as an input. Again, not your fault, something to fix in Rust also, and it should implement the following logic: https://github.com/consensus-shipyard/ipc-actors/blob/2d2428eb37edc83155d8706e5c5f8ddb1f526ec8/sdk/src/subnet_id.rs#L187

Reply:
Done